### PR TITLE
Modify ossl_provider_forall_loaded() to avoid locking for the callbacks

### LINKED ISSUE #14251: ossl_provider_forall_loaded should not hold a lock while calling the user callback
A user callback could do almost anything, so a deadlock is a very likely outcome.

This is very similar to the issue encountered in #14225 and fixed by #14250

### PR DESCRIPTION
**Pull request by paulidale**
This avoids recursive lock issues by taking a copy of the provider list and he callbacks are made without holding the store lock.

Fixes #14251

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by slontis — COMMENTED** [review]
update the second commit to not say 'atatement'..

**Comment by slontis** [comment]
Commit should also say Fixes #??

File: doc/internal/man3/ossl_provider_new.pod [hunk 1]
```diff
@@ -199,8 +199,6 @@ passed back to some of the provider functions.
 
 ossl_provider_forall_loaded() iterates over all the currently
 "activated" providers, and calls I<cb> for each of them.
-If no providers have been "activated" yet, it tries to activate all
```
Comment (slontis) [inline doc/internal/man3/ossl_provider_new.pod:202]: Why was this removed??

File: crypto/provider_core.c [hunk 2]
```diff
@@ -824,18 +795,46 @@ int ossl_provider_forall_loaded(OSSL_LIB_CTX *ctx,
...
+
+        if (prov->flag_activated && !cb(prov, cbdata))
+            goto finish;
+    }
+
+    ret = 1;
+    goto finish;
+
+ err_unlock:
+    CRYPTO_THREAD_unlock(store->lock);
```
Comment (slontis) [inline crypto/provider_core.c:832]: Is i correct in the case where it does the goto err_unlock?

File: crypto/provider_core.c [hunk 3]
```diff
@@ -824,18 +795,46 @@ int ossl_provider_forall_loaded(OSSL_LIB_CTX *ctx,
...
+            goto err_unlock;
+    CRYPTO_THREAD_unlock(store->lock);
 
+    /*
+     * Now, we sweep through all providers not under lock
+     */
+    for (j = 0; j < i; j++) {
+        OSSL_PROVIDER *prov = sk_OSSL_PROVIDER_value(provs, j);
+
+        if (prov->flag_activated && !cb(prov, cbdata))
```
Comment (slontis) [inline crypto/provider_core.c:824]: Is the cb ever likely to do ossl_provider_free()? What happens in that scenario?

File: doc/internal/man3/ossl_provider_new.pod (hunk 1 above)
Comment (paulidale) [reply doc/internal/man3/ossl_provider_new.pod:202]: Because it is a lie.  The code doesn't do this.

I've no idea why the code doesn't do it.

File: crypto/provider_core.c (hunk 2 above)
Comment (paulidale) [reply crypto/provider_core.c:832]: Yes, that's why the loop variables get swapped.

File: crypto/provider_core.c (hunk 3 above)
Comment (paulidale) [reply crypto/provider_core.c:824]: If the callback _owns_ a reference to the provider, it is entitled to free it.  It shouldn't free the reference that this function has --> it doesn't own it.  If the callback frees the provider (and no other references exist), the provider will persist until this function frees it later.

This isn't a case I'm too concerned about, the ref counting is more to avoid threading issues.

It would be possible to more eagerly free the local references to providers, but it doesn't seem all that useful.

**Comment by paulidale** [comment]
The first commit includes a _Fixes_ line.

File: crypto/provider_core.c (hunk 3 above)
Comment (slontis) [reply crypto/provider_core.c:824]: I was thinking of the case where the for_all is used to perform a free of its objects..

Comment (paulidale) [reply crypto/provider_core.c:824]: That works.  They hang around until this function frees them after the callbacks are all done.

**Comment by slontis** [comment]
> The first commit includes a _Fixes_ line.

Added to the PR comment above also

**Comment by paulidale** [comment]
@levitte, any reason this function is `forall` while all the rest are `doall`?

**Comment by paulidale** [comment]
The suggestion is to rename this `ossl_provider_doall_activated()`

File: doc/internal/man3/ossl_provider_new.pod (hunk 1 above)
Comment (mattcaswell) [reply doc/internal/man3/ossl_provider_new.pod:202]: Isn't that what the `provider_activate_fallbacks` call does?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/provider_core.c [hunk 4]
```diff
@@ -824,18 +795,46 @@ int ossl_provider_forall_loaded(OSSL_LIB_CTX *ctx,
...
+            goto finish;
+    }
+
+    ret = 1;
+    goto finish;
+
+ err_unlock:
+    CRYPTO_THREAD_unlock(store->lock);
+ finish:
+    /* The pop_free call doesn't do what we want on an error condition */
```
[the remark this answers is not in the document]
Comment (paulidale) [reply crypto/provider_core.c:834]: Not all elements have had their reference count incremented on the error path.  Pop_free always processes all elements.

**Comment by paulidale** [comment]
Feedback addressed again.

**Review by slontis — APPROVED** [review]
LGTM..

The concerning thing here is the test failure in what seems like the rsa keygen..
I am sticking this is a loop to see if I can get it to happen again... Although I have not seen this error before and that code has been in for a very long time now..

**Comment by paulidale** [comment]
It's a heisen-bug :(

**Comment by slontis** [comment]
> It's a heisen-bug :(

I cant be certain :)

I have been running that line in a loop for an hour with no issues.

**Comment by paulidale** [comment]
My guess would be that it was related to the random seed being used.  RSA keygen can fail if it takes too long.  It looked like long output.

**Comment by paulidale** [comment]
Merged to master, thanks for the feedbacks.
